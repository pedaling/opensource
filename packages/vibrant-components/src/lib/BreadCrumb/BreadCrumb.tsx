import { Link } from '@vibrant-ui/core';
import { Body } from '../Body';
import { useBreadCrumb } from '../BreadCrumbProvider';
import { Pressable } from '../Pressable';
import { withBreadCrumbVariation } from './BreadCrumbProps';

export const BreadCrumb = withBreadCrumbVariation(({ children, href, isExternal, onClick }) => {
  const { current } = useBreadCrumb();

  return (
    <>
      {href ? (
        <Link href={href} isExternal={isExternal} onClick={onClick}>
          <Body level={2} weight="regular" color={current ? 'onView1' : 'onView2'} lineLimit={1}>
            {children}
          </Body>
        </Link>
      ) : (
        <Pressable onClick={onClick} cursor={onClick ? 'pointer' : 'default'}>
          <Body level={2} weight="regular" color={current ? 'onView1' : 'onView2'} lineLimit={1}>
            {children}
          </Body>
        </Pressable>
      )}
    </>
  );
});
