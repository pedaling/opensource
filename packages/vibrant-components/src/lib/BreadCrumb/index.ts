export { BreadCrumb } from './BreadCrumb';
export type { BreadCrumbProps } from './BreadCrumbProps';
