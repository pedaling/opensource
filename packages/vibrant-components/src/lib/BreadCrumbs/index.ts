export { BreadCrumbs } from './BreadCrumbs';
export type { AriaCurrent } from './BreadCrumbsProps';
