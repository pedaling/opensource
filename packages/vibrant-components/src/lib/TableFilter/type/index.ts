export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
