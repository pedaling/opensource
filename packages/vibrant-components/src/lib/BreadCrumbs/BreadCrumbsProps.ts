import type { ComponentType, ReactElement, Ref } from 'react';
import { withVariation } from '@vibrant-ui/core';
import type { IconProps } from '@vibrant-ui/icons';
import type { BreadCrumbProps } from '../BreadCrumb';

export type BreadCrumbsProps = {
  children: ReactElement<BreadCrumbProps> | ReactElement<BreadCrumbProps>[];
  Separator: ComponentType<IconProps> | string;
  ref?: Ref<any>;
};

export const withBreadCrumbsVariation = withVariation<BreadCrumbsProps>('BreadCrumbs')();
