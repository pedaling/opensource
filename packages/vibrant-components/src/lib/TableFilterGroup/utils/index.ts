import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
