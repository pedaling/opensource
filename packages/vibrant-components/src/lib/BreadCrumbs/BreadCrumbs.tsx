/* eslint-disable @typescript-eslint/naming-convention */
import { Children, isValidElement } from 'react';
import { Box } from '@vibrant-ui/core';
import { Icon } from '@vibrant-ui/icons';
import { Body } from '../Body';
import type { BreadCrumbProps } from '../BreadCrumb';
import { BreadCrumbProvider } from '../BreadCrumbProvider';
import { HStack } from '../HStack';
import { Space } from '../Space';
import { withBreadCrumbsVariation } from './BreadCrumbsProps';

export const BreadCrumbs = withBreadCrumbsVariation(
  ({ innerRef, children, Separator = Icon.ChevronRight.Thin, ariaCurrent = 'true' }) => (
    <HStack as="nav" ariaLabel="Breadcrumb">
      <HStack as="ol" alignVertical="center" flexWrap="wrap" ref={innerRef}>
        {Children.toArray(children)
          .filter(isValidElement<BreadCrumbProps>)
          .map((child, index, array) =>
            index === 0 ? (
              <HStack as="li" key={`child-${index}`}>
                {child}
                {array.length > 1 && <Space width={4} />}
              </HStack>
            ) : (
              <Box
                as="li"
                key={`child-${index}`}
                flexDirection="row"
                alignItems="center"
                ariaCurrent={index === array.length - 1 ? ariaCurrent : undefined}
              >
                <Box>
                  {typeof Separator === 'string' ? (
                    <Body level={2} weight="regular" color="onView2">
                      {Separator}
                    </Body>
                  ) : (
                    <Separator fill="onView2" size={12} />
                  )}
                </Box>
                <Space width={4} />
                {index === array.length - 1 ? (
                  <BreadCrumbProvider current={true}>
                    <>{child}</>
                  </BreadCrumbProvider>
                ) : (
                  child
                )}
                {index < array.length - 1 ? <Space width={4} /> : null}
              </Box>
            )
          )}
      </HStack>
    </HStack>
  )
);
