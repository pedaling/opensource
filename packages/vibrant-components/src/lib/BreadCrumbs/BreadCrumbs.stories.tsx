import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { action } from '@vibrant-ui/utils';
import { BreadCrumb } from '../BreadCrumb';
import { VStack } from '../VStack';
import { BreadCrumbs } from './BreadCrumbs';

export default {
  title: 'BreadCrumbs',
  component: BreadCrumbs,
  args: {},
} as ComponentMeta<typeof BreadCrumbs>;

export const Basic: ComponentStory<typeof BreadCrumbs> = props => (
  <VStack p={10}>
    <BreadCrumbs {...props}>
      <BreadCrumb href="/">Depth1 has link</BreadCrumb>
      <BreadCrumb href="/" isExternal={true}>
        Depth2 has external link
      </BreadCrumb>
      <BreadCrumb onClick={action('Depth3 is clicked')}>Depth3 has onClick function</BreadCrumb>
      <BreadCrumb>Last depth is highlighted</BreadCrumb>
    </BreadCrumbs>
  </VStack>
);

export const withStringSeparator: ComponentStory<typeof BreadCrumbs> = props => (
  <VStack p={10}>
    <BreadCrumbs Separator="+" {...props}>
      <BreadCrumb href="/">Depth1 has link</BreadCrumb>
      <BreadCrumb href="/" isExternal={true}>
        Depth2 has external link
      </BreadCrumb>
      <BreadCrumb onClick={action('Depth3 is clicked')}>Depth3 has onClick function</BreadCrumb>
      <BreadCrumb>Last depth is highlighted</BreadCrumb>
    </BreadCrumbs>
  </VStack>
);

export const withLongName: ComponentStory<typeof BreadCrumbs> = props => (
  <VStack p={10}>
    <BreadCrumbs {...props}>
      <BreadCrumb>{'Depth1 has long name '.repeat(3)}</BreadCrumb>
      <BreadCrumb>{'Depth2 has long name '.repeat(5)}</BreadCrumb>
      <BreadCrumb>{'Depth3 has long name '.repeat(30)}</BreadCrumb>
      <BreadCrumb>{'Depth4 has long name '.repeat(30)}</BreadCrumb>
    </BreadCrumbs>
  </VStack>
);
