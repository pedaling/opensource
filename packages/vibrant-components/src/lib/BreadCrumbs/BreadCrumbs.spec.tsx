import { fireEvent } from '@testing-library/react';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { lightModeColors } from '../../../../vibrant-theme/src/lib/theme/lightModeColors';
import { BreadCrumb } from '../BreadCrumb';
import { BreadCrumbs } from './BreadCrumbs';

describe('<BreadCrumbs />', () => {
  const { render } = createReactRenderer();

  let renderer: ReactRenderer;
  let mockOnClick: jest.Mock;
  let breadcrumbTexts: HTMLElement[];

  beforeEach(() => {
    mockOnClick = jest.fn();
  });

  describe('when BreadCrumbs with 3 items rendered', () => {
    beforeEach(async () => {
      renderer = render(
        <BreadCrumbs testId="BreadCrumbs">
          <BreadCrumb testId="BreadCrumb">BreadCrumb1</BreadCrumb>
          <BreadCrumb testId="BreadCrumb">BreadCrumb2</BreadCrumb>
          <BreadCrumb testId="BreadCrumb">BreadCrumb3</BreadCrumb>
        </BreadCrumbs>
      );

      breadcrumbTexts = await renderer.findAllByTestId('breadcrumb-text');
    });

    it('should render 3 BreadCrumb', async () => {
      expect(renderer.getAllByTestId('BreadCrumb').length).toBe(3);
    });

    it('should render 2 separators', async () => {
      expect(renderer.getAllByTestId('separator').length).toBe(2);
    });

    it('non last BreadCrumb should have onView2 color', () => {
      breadcrumbTexts.forEach((text, idx, arr) => {
        if (idx === arr.length - 1) return;

        expect(text).toHaveStyleRule('color', lightModeColors.onView2);
      });
    });

    it('last BreadCrumb should have onView1 color', () => {
      expect(breadcrumbTexts[breadcrumbTexts.length - 1]).toHaveStyleRule('color', lightModeColors.onView1);
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when separator is provided at BreadCrumbs', () => {
    beforeEach(() => {
      renderer = render(
        <BreadCrumbs testId="BreadCrumbs" Separator="*">
          <BreadCrumb testId="BreadCrumb">BreadCrumb1</BreadCrumb>
          <BreadCrumb testId="BreadCrumb">BreadCrumb2</BreadCrumb>
        </BreadCrumbs>
      );
    });

    it('should render separator with provided text', () => {
      expect(renderer.getByTestId('separator').textContent).toBe('*');
    });
  });

  describe('when href provided at BreadCrumb', () => {
    beforeEach(() => {
      renderer = render(
        <BreadCrumbs testId="BreadCrumbs">
          <BreadCrumb testId="BreadCrumb" href="https://www.vibrant-design.com">
            BreadCrumb1
          </BreadCrumb>
        </BreadCrumbs>
      );
    });

    it('should render link with href attribute', () => {
      expect(renderer.getByRole('link').getAttribute('href')).toBe('https://www.vibrant-design.com');
    });
  });

  describe('when isExternal is true at BreadCrumb', () => {
    beforeEach(() => {
      renderer = render(
        <BreadCrumbs testId="BreadCrumbs">
          <BreadCrumb testId="BreadCrumb" href="https://www.vibrant-design.com" isExternal={true}>
            BreadCrumb1
          </BreadCrumb>
        </BreadCrumbs>
      );
    });

    it('should open link with new tab', () => {
      expect(renderer.getByRole('link').getAttribute('target')).toBe('_blank');
    });
  });

  describe('when onClick provided at BreadCrumb and is clicked', () => {
    beforeEach(() => {
      renderer = render(
        <BreadCrumbs testId="BreadCrumbs">
          <BreadCrumb testId="BreadCrumb" onClick={mockOnClick}>
            BreadCrumb1
          </BreadCrumb>
        </BreadCrumbs>
      );

      fireEvent.click(renderer.getByTestId('BreadCrumb'));
    });

    it('should call onClick', () => {
      expect(mockOnClick).toHaveBeenCalled();
    });
  });
});
