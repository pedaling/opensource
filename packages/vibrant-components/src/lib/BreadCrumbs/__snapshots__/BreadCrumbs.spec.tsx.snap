// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreadCrumbs /> when BreadCrumbs with 3 items rendered match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: relative;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: relative;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  cursor: default;
  background-color: inherit;
  border-width: 0;
  padding: 0;
  overflow: hidden;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  z-index: 0;
}

.emotion-4 {
  text-align: left;
  color: #373737cc;
  -webkit-text-fill-color: #373737cc;
  text-fill-color: #373737cc;
  display: -webkit-box;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 1;
  font-size: 0.875rem;
  line-height: 1.125rem;
  font-weight: 400;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 4px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  position: relative;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 12px;
  height: 12px;
  fill: #373737cc;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
}

.emotion-20 {
  text-align: left;
  color: #0c0c0c;
  -webkit-text-fill-color: #0c0c0c;
  text-fill-color: #0c0c0c;
  display: -webkit-box;
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 1;
  font-size: 0.875rem;
  line-height: 1.125rem;
  font-weight: 400;
}

<div>
  <nav
    aria-label="Breadcrumb"
    class="emotion-0"
    data-testid="BreadCrumbs"
  >
    <ol
      class="emotion-1"
    >
      <li
        class="emotion-0"
      >
        <button
          class="emotion-3"
          data-testid="BreadCrumb"
          style="opacity: 1; transition-property: opacity; transition-duration: 200ms; transition-timing-function: cubic-bezier(0.5, 1, 0.89, 1);"
          type="button"
        >
          <span
            class="emotion-4"
            data-testid="breadcrumb-text"
          >
            BreadCrumb1
          </span>
        </button>
        <div
          class="emotion-5"
        />
      </li>
      <li
        class="emotion-6"
      >
        <div
          class="emotion-0"
        >
          <svg
            class="emotion-8"
            data-testid="separator"
            viewBox="0 0 24 24"
          >
            <path
              class="emotion-9"
              d="m15.955 11.525-8.2 8.2c-.1.1-.1.25 0 .35l.9.9c.1.1.25.1.35 0l9.25-9.25c.1-.1.1-.25 0-.35l-9.25-9.3c-.1-.1-.25-.1-.35 0l-.9.9c-.1.1-.1.25 0 .35z"
            />
          </svg>
        </div>
        <div
          class="emotion-5"
        />
        <button
          class="emotion-3"
          data-testid="BreadCrumb"
          style="opacity: 1; transition-property: opacity; transition-duration: 200ms; transition-timing-function: cubic-bezier(0.5, 1, 0.89, 1);"
          type="button"
        >
          <span
            class="emotion-4"
            data-testid="breadcrumb-text"
          >
            BreadCrumb2
          </span>
        </button>
        <div
          class="emotion-5"
        />
      </li>
      <li
        class="emotion-6"
      >
        <div
          class="emotion-0"
        >
          <svg
            class="emotion-8"
            data-testid="separator"
            viewBox="0 0 24 24"
          >
            <path
              class="emotion-9"
              d="m15.955 11.525-8.2 8.2c-.1.1-.1.25 0 .35l.9.9c.1.1.25.1.35 0l9.25-9.25c.1-.1.1-.25 0-.35l-9.25-9.3c-.1-.1-.25-.1-.35 0l-.9.9c-.1.1-.1.25 0 .35z"
            />
          </svg>
        </div>
        <div
          class="emotion-5"
        />
        <button
          class="emotion-3"
          data-testid="BreadCrumb"
          style="opacity: 1; transition-property: opacity; transition-duration: 200ms; transition-timing-function: cubic-bezier(0.5, 1, 0.89, 1);"
          type="button"
        >
          <span
            class="emotion-20"
            data-testid="breadcrumb-text"
          >
            BreadCrumb3
          </span>
        </button>
      </li>
    </ol>
  </nav>
</div>
`;
